### ISSUE TITLE
[RFC] Support nullability for jax Array

### ISSUE THREAD

**Issue by chaokunyang** [google/jax#16289]
# Motivation
Nullability is supported in apache arrow, when converting arrow int array with nulls to numpy or jax, the type changed from `pa.int8/16/32/64` to float automatically to use `np.nan` to express null. It's OK for `int8/int16/int32` since it can be expressed using float64 precision bits, but when it comes to int64, converting from int64 to flaot64 will loss  precision. 

Since jax is used for numeric computing widely, It would be great to support nullability natively in jax. For example, we can use a bitmap to expression nullability for every element in array, and propagate the nullability along the computing graph.

It may be not easy, since the XLA is developed for deep learning workload, and the int64 precision is not a priority.

Without the nullability support in jax, we have to define a higher-level expression on jax to propagate the nullability using a bool array.

# Solution1
- Add a bitmap region for every jax array.
- Add extra XLA computing for bitmap along all jax arrays 

# Solution2
Solution1 may introduce extra cost for deeplearning since it doesn't have strict precision requirements.

To avoid introduce breaking changes or performance cost, we can introduce a new `jax.Array` type, i.e. `jax.NullableArray` , which carries a bitmap, and make all computing with jax.NullableArray` to progate the bitmaps: the result bitmaps are union of every bitmap

# Checks

- [x] Check for duplicate requests.
- [ ] Describe your goal, and if possible provide a code snippet with a motivating example.

**Comment by jakevdp** [comment]
Thanks for the request! JAX previously included an experimental masking transformation that accomplished something like this, but it did not get enough use and was eventually deprecated and removed.
